### ISSUE TITLE
Bug in the backend

### ISSUE THREAD

**Issue by jotpe** [drolbr/Overpass-API#201]
For a little project I need the Overpass API to run with 64 bit integers. But looking at the current node ids with values about 3.3 billion: the overflow will come.

Regards Johannes

**Comment by drolbr** [comment]
Overpass API uses 64-bit-integers for node ids since version 0.7. Hence 
there will be no overflow.

**Comment by jotpe** [comment]
Wise! 

But what could this mean?

>bin/update_database --db-dir=/tmp/db --meta < ~/workspace/overpass-wiki/commons_all2.xml
Reading XML file ... elapsed node 1336995620. Flushing to database ...... done.
Reading XML file ... finished reading nodes. Flushing to database ...... done.
Reorganizing the database ...File error caught: 2 No such file or directory /tmp/db/nodes.0a.bin Block_Backend: index out of range.

**Comment by drolbr** [comment]
This is a message from the file backend.

Every "File_Error" is emitted somewhere in the backend. That 
"Block_Backend" reports an "index out of range" means that the data in 
the index file ".idx" is inconsistent with the actual data in the data 
file ".bin".

Most likely this is because a concurrent process has written to one of 
the involved files. Please check whether such a process might have been 
running.

The second biggest probability is a bug in the program. If this is the 
case then most likely re-running the program will reporduce exactly the 
same error message.

The third most likely probability is a damaged harddisk. In this case, 
re-running will most likely produce a similar, but not exactly equal 
error message.

**Comment by jotpe** [comment]
Rerunning with a fresh db directory leads to exactly the same message.

**Comment by mmd-osm** [comment]
@jotpe : is your file `commons_all2.xml` available for testing somewhere?

**Comment by jotpe** [comment]
Yes, [here](https://drive.google.com/folderview?id=0B3PppnqNatwyfjNHbDJibFJ4YU9zb0tFemNYamZxUm9KS216V29oMGZEeUp2bFVnLWt4U1k&usp=sharing).

**Comment by mmd-osm** [comment]
Thanks for providing a sample file! I couldn't really reproduce the issue on my branch test753 and I would assume this to also be the cast with the current master. Which branch did you use for your tests?

Another point: I've noticed that you provided a `--db-dir=` parameter for update_database. Did you also run a dispatcher at the same time?

Here's my set up:

Step 1: Create a new directory osm3s_test.

Step 2: Start dispatcher
```
./dispatcher --osm-base --attic --db-dir=/home/{{user}}/osm3s_test
```

Step 2: Start Import

```
cat commons_all2.xml | ./update_database --meta
Reading XML file ... elapsed node 1336995620. Flushing to database ...... done.
Reading XML file ... finished reading nodes. Flushing to database ...... done.
Flushing to database ...... done.
Flushing to database ....... done.
Update complete.
```

**Comment by jotpe** [comment]
I used the issue_178 branch.

I've 8G available. The update process can run without swapping, if every user application is closed.

No, I did not run the dispatcher, while executing update_database.

Following your steps, no error appears.

**Comment by jotpe** [comment]
This successful run does not print: Reorganizing the database.

>bin/update_database  --meta < ~/workspace/overpass-wiki/commons_all2.xml 
Reading XML file ... elapsed node 1336995620. Flushing to database ...... done.
Reading XML file ... finished reading nodes. Flushing to database ...... done.
Flushing to database ...... done.
Flushing to database ....... done.
Update complete.

**Comment by mmd-osm** [comment]
ok... cross checking your variant: issue is reproducible. 

```
 ./update_database --db-dir=/home/{{user}}/osm3s_test --meta < commons_all2.xml 
Reading XML file ... elapsed node 1336995620. Flushing to database ...... done.
Reading XML file ... finished reading nodes. Flushing to database ...... done.
Reorganizing the database ...File error caught: 2 No such file or directory /home/{{user}}/osm3s_test/nodes.0a.bin Block_Backend: index out of range.
```

Obviously, the non-transactional update behaves differently with the additional reorganizing step causes some issue.

**Comment by mmd-osm** [comment]
Looking at `init_osm3s.sh`, your variant (without dispatcher) seems to be the correct one.

I just started a planet import via dispatcher. The missing *Reorganizing the database* step leads to quite huge files very early during the file processing . My conclusion here is that for a planet, update_database via dispatcher doesn't seem to be feasible.

In the case of your rather small input file, this missing reorganization step didn't seem to make a large difference, though.

**Comment by mmd-osm** [comment]
A workaround is to set flush-size to 0, forcing processing of the whole file at once. 

```
./update_database --db-dir=/home/user/db/ --meta --flush-size=0 < /home/user/commons_all2.xml
Reading XML file ... finished reading nodes. Flushing to database ...... done.
Flushing to database ...... done.
Flushing to database ....... done.
Update complete.
```

Besides, nodes.0a.bin is already corrupt before it hits the merge_files step.

**Comment by mmd-osm** [comment]
Honestly, the test data you provided is really crappy. It includes corrupted timestamps, missing user names and most disturbingly lat / lon values, which are outright invalid. I should have checked this right from the beginning... :camel: 

```
sed -e "s/user='' uid=''/user='unknown' uid='999999'/g"                       \
    -e "s/timestamp='2014-12-03T09'/timestamp='2014-12-03T09:00:00Z'/g"       \
    -e "s/timestamp='2010-12-29T18:35:5'/timestamp='2010-12-29T18:35:50Z'/g"  \
    -e "s/timestamp='2014-07-'/timestamp='2014-07-01T00:00:00Z'/g"            \
    -e "s/timestamp='2014-06-21T22'/timestamp='2014-06-21T22:00:00Z'/g"       \
    -e "s/timestamp='2013-09-25T10:'/timestamp='2013-09-25T10:00:00Z'/g"      \
    -e "s/timestamp='2011-'/timestamp='2011-01-01T00:00:00Z'/g"               \
    -e "s/timestamp='2014-01-21T16:'/timestamp='2014-01-21T16:00:00Z'/g"      \
    -e "s/timestamp='2013-03-'/timestamp='2013-03-01T00:00:00Z'/g"            \
    -e "s/timestamp='2013-10-17T16:44:4'/timestamp='2013-10-17T16:44:44Z'/g"  \
    -e "s/timestamp='2014-08-11T'/timestamp='2014-08-11T00:00:00Z'/g"         \
    -e "s/timestamp='2013-03-2'/timestamp='2013-03-20T00:00:00Z'/g"           \
    -e "s/lat='5595275.0'/lat='55.952750'/g"                                  \
    -e "s/lon='449867.0'/lon='44.98670'/g"                                    \
    -e "s/lon='692781.0'/lon='69.27810'/g"                                    \
     commons_all2.xml 
```

Once those issues are fixed via the script above, `update_database`  finishes without issue:

```
./update_database --db-dir=db --compression_method=no < ~/jotpe/c2.xml 
Reading XML file ... elapsed node 1336995620. Flushing to database ...... done.
Reading XML file ... finished reading nodes. Flushing to database ...... done.
Reorganizing the database ... done.
Flushing to database ...... done.
Flushing to database ....... done.
Update complete.
```

TODO for `update_database`:

* Check if lat / lon values are valid

**Comment by drolbr** [comment]
Lat, lon: fixed in commit 7f086efce2b820210e66c0fa1565613fbfad8452

The absence of user name and user id and weird timestamps do not hurt.
